kuma/javascript/src/document-provider.jsx

```diff
@@ -51,20 +51,32 @@ export default function DocumentProvider(
             throw new Error('DocumentProvider effect ran without body.');
         }
         const body = document.body;
+        const fetchConfig = { redirect: 'follow' };
+
+        function getJSON(response) {
+            if (response.ok) {
+                return response.json();
+            } else {
+                // If we didn't get a good response, throw an error.
+                throw new Error(`${response.status}:${response.statusText}`);
+            }
+        }
 
         // This is the function that does client side navigation
         function navigate(url, slug) {
             body.style.opacity = '0.15';
-            fetch(`/api/v1/doc/${locale}/${slug}`, { redirect: 'follow' })
+            fetch(`/api/v1/doc/${locale}/${slug}`, fetchConfig)
                 .then(response => {
-                    if (response.ok) {
-                        return response.json();
+                    if (locale !== 'en-US' && response.status === 404) {
+                        // We requested a non-English document that didn't
+                        // exist, but before we abandon client-side navigation,
+                        // let's first try falling-back to the English document.
+                        return fetch(
+                            `/api/v1/doc/en-US/${slug}`,
+                            fetchConfig
+                        ).then(getJSON);
```

Comment:
JavaScript promises are awesome when they work well, but it can be really hard sometimes to figure out how to make them work well.  Ideally, all the .then() calls are in a single linear chain, and having to use a nested one like this is a code smell.

The code seems fine, and I'm not opposed to landing it this way. But factoring out the getJSON function doesn't seem ideal to me.

I think that the better approach might be to write a utility function fetchWithFallback() that can take one or two urls. With one url, it works just like fetch(). But with two URLs, if it gets a 404 on the first, it fetches the second.  If you had such a utility function, then this change would fit trivially into the existing promise chain and all you'd have to do is the logic of deciding to pass one url (when the locale is en-US) or two urls.

Reply:
I'm going to get some coffee and then try to write this utility function.

Reply:
This works for me:

```
function fetchWithFallback(url1: string, url2: string) {
  const fetchConfig = { redirect: "follow" };
  if (!url2 || url2 === url1) {
    return fetch(url1, fetchConfig);
  }

  return fetch(url1, fetchConfig).then(response => {
    if (response.ok) {
      return response;
    } else {
      return fetch(url2, fetchConfig);
    }
  });
}
```

Reply:
Actually, this is a more compact version that also seems to work:

```
function fetchWithFallback(url1: string, url2: string) {
  const fetchConfig = { redirect: "follow" };
  return fetch(url1, fetchConfig).then(response => {
    if (response.ok || !url2 || url2 === url1) {
      return response;
    } else {
      return fetch(url2, fetchConfig);
    }
  });
}
```

Reply:
Note that the way this function compares the two urls, you could simply call it with the original url and the same url forced to en-US and wouldn't even have to care about whether they were actually the same url.

Reply:
Nice, I like it. I'll make that change, thanks @davidflanagan!